File: CLI/JsonCommand.cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Core.Models;
using Services.Concrete;
using Services.Concrete.Analyzer;

namespace CLI
{
    internal class JsonCommand : ExportCommand
    {
        public const string COMMAND_NAME = "json";

        public JsonCommand() : base(COMMAND_NAME, @"Export demos into JSON files")
        {
        }

        public override void PrintHelp()
        {
            Console.WriteLine(GetDescription());
            Console.WriteLine(@"");
            Console.WriteLine($@"Usage: {Program.ExeName} {COMMAND_NAME} demoPaths... [--output] [--source] [--force-analyze]");
            Console.WriteLine(@"");
            Console.WriteLine(@"Demos path can be either .dem files location or a directory. It can be relative or absolute.");
            Console.WriteLine(@"The --output argument specify the directory where output files will be saved.");
            Console.WriteLine($@"The --source argument force the analysis logic of the demo analyzer. Available values: [{string.Join(",", _availableSources)}]");
            Console.WriteLine(@"The --force-analyze argument force demos analyzes (ignore cached data).");
            Console.WriteLine(@"");
            Console.WriteLine(@"Examples:");
            Console.WriteLine(@"");
            Console.WriteLine(@"Export a demo:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\demo.dem""");
            Console.WriteLine(@"");
            Console.WriteLine(@"Export multiple demos:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\demo.dem"" ""C:\Users\username\Desktop\demo2.dem""");
            Console.WriteLine(@"");
            Console.WriteLine(@"Export all demos in a directory using the ESL analyzer and save it in a custom directory:");
            Console.WriteLine($@"    {Program.ExeName} {COMMAND_NAME} ""C:\Users\username\Desktop\MyFolder"" --output ""C:\Users\username\Documents"" --source esl");
        }

        private new void ParseArgs(string[] args)
        {
            base.ParseArgs(args, new string[] { });
        }

        public override async Task Run(string[] args)
        {
            ParseArgs(args);

            if (_demoPaths.Count == 0)
            {
                Console.WriteLine(@"No demo paths provided");
                return;
            }

            try
            {
                CacheService cacheService = new CacheService();
                int currentDemoNumber = 0;
                foreach (string demoPath in _demoPaths)
                {
                    Console.WriteLine($@"Retrieving demo {++currentDemoNumber}/{_demoPaths.Count} {demoPath}");
                    Demo demo = DemoAnalyzer.ParseDemoHeader(demoPath);
                    if (demo == null)
                    {
                        Console.WriteLine($@"Invalid demo {demoPath}");
                        continue;
                    }

                    if (_source != null)
                    {
                        demo.Source = _source;
                    }

                    if (!_forceAnalyze && cacheService.HasDemoInCache(demo.Id))
                    {
                        demo = await cacheService.GetDemoDataFromCache(demo.Id);
                        demo.WeaponFired = await cacheService.GetDemoWeaponFiredAsync(demo);
                        demo.PlayerBlinded = await cacheService.GetDemoPlayerBlindedAsync(demo);
                    }
                    else
                    {
                        try
                        {
                            Console.WriteLine($@"Analyzing demo {demoPath}");
                            DemoAnalyzer analyzer = DemoAnalyzer.Factory(demo);
                            demo = await analyzer.AnalyzeDemoAsync(new CancellationTokenSource().Token);
                            await cacheService.WriteDemoDataCache(demo);
                        }
                        catch (Exception)
                        {
                            Console.WriteLine($@"Error while analyzing demo {demoPath}");
                            continue;
                        }
                    }

                    string outputFolderPath = BuildOutputFolderPathFromDemoPath(demoPath);
                    string jsonFilePath = await cacheService.GenerateJsonAsync(demo, outputFolderPath);
                    Console.WriteLine($@"JSON file generated at {jsonFilePath}");
                }
            }
            catch (FileNotFoundException ex)
            {
                Console.WriteLine($@"The demo doesn't exists: {ex.FileName}");
                Environment.Exit(1);
            }
            catch (Exception ex)
            {
                Console.WriteLine($@"Error while exporting demo: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: src/cli/commands/json-command.ts
import { SupportedDemoSources } from 'csdm/common/types/counter-strike';
import { exportMatchesToJson } from 'csdm/node/json/export-matches-to-json';
import { ExportCommand } from './export-command';

export class JsonCommand extends ExportCommand {
  public static Name = 'json';
  private minify = false;
  private minifyFlag = '--minify';

  public getDescription() {
    return 'Export demos into JSON files.';
  }

  public printHelp() {
    const sourceValues = SupportedDemoSources.join(',');
    console.log(this.getDescription());
    console.log('');
    console.log(
      `Usage: csdm ${JsonCommand.Name} demoPaths... ${this.formatFlagsForHelp(this.commonFlags)} ${this.formatFlagForHelp(this.minifyFlag)}`,
    );
    console.log('');
    console.log('Demos path can be either a .dem files path or a directory. It can be relative or absolute.');
    console.log(`The ${this.outputFolderFlag} flag specify the directory where output files will be saved.`);
    console.log(
      `The ${this.sourceFlag} flag forces the analysis logic of the demo analyzer. Available values: [${sourceValues}]`,
    );
    console.log(`The ${this.forceAnalyzeFlag} flag force demos analyzes even if they are already in the database.`);
    console.log('');
    console.log('Examples:');
    console.log('');
    console.log('Export 1 demo into a minified JSON file:');
    console.log(`    csdm ${JsonCommand.Name} "C:\\Users\\username\\Desktop\\demo.dem" ${this.minifyFlag}`);
    console.log('');
    console.log('Export all demos in a directory using the ESL analyzer and save it in a custom directory:');
    console.log(
      `    csdm ${JsonCommand.Name} "C:\\Users\\username\\Desktop\\MyFolder" ${this.outputFolderFlag} "C:\\Users\\username\\Documents" ${this.sourceFlag} esl`,
    );
  }

  public async run() {
    await this.parseArgs();
    await super.run();

    const checksums = this.demosToExport.map((demo) => demo.checksum);
    const options = {
      outputFolderPath: this.outputFolderPath,
      checksums,
      minify: this.minify,
    };
    try {
      await exportMatchesToJson(options);
      console.log(`Exported ${this.demosToExport.length} matches in folder ${this.outputFolderPath}`);
    } catch (error) {
      console.log('Error while exporting matches to JSON');
    }
  }

  protected async parseArgs() {
    await super.parseArgs();

    for (let index = 0; index < this.args.length; index++) {
      const arg = this.args[index];
      if (!this.isFlagArgument(arg)) {
        continue;
      }

      switch (arg) {
        case this.minifyFlag:
          this.minify = true;
          break;
        default:
          if (!super.isCommonFlag(arg)) {
            console.log(`Unknown flag: ${arg}`);
            this.exitWithFailure();
          }
      }
    }
  }
}
